File: lib/python/openlinktoken/src/test/openlinktoken/tokentransformer/jwe_match_token_formatter_test.py
# SPDX-License-Identifier: MIT
"""
Unit tests for JweMatchTokenFormatter.
"""

import base64
import json
import unittest

from jwcrypto import jwe, jwk

from openlinktoken.crypto_suite import CryptoSuite
from openlinktoken.tokentransformer.jwe_match_token_formatter import JweMatchTokenFormatter


class TestJweMatchTokenFormatter(unittest.TestCase):
    """Unit tests for JweMatchTokenFormatter."""

    TEST_ENCRYPTION_KEY = "12345678901234567890123456789012"  # 32 chars
    TEST_RING_ID = "test-ring-2026"
    TEST_RULE_ID = "T1"
    TEST_TOKEN = "dGVzdC10b2tlbi1wcGlk"  # base64-encoded test token
    TEST_ISSUER = "test.issuer"

    def test_constructor_with_valid_parameters(self):
        """Test that constructor works with valid parameters."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER
        )
        self.assertIsNotNone(formatter)
        self.assertEqual(formatter.ring_id, self.TEST_RING_ID)
        self.assertEqual(formatter.rule_id, self.TEST_RULE_ID)
        self.assertEqual(formatter.issuer, self.TEST_ISSUER)

    def test_constructor_with_null_encryption_key(self):
        """Test that constructor raises error with null encryption key."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter(None, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER)
        self.assertIn("32 bytes", str(context.exception))

    def test_constructor_with_invalid_key_length(self):
        """Test that constructor raises error with invalid key length."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter("short", self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER)
        self.assertIn("32 bytes", str(context.exception))

    def test_constructor_rejects_non_ascii_string_with_32_characters_but_more_than_32_bytes(self):
        """UTF-8 multi-byte key strings must be rejected before JWE setup."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter("é" * 32, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER)
        self.assertIn("32 bytes", str(context.exception))

    def test_constructor_with_null_ring_id(self):
        """Test that constructor raises error with null ring ID."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, None, self.TEST_RULE_ID, self.TEST_ISSUER)
        self.assertIn("Ring ID", str(context.exception))

    def test_constructor_with_empty_ring_id(self):
        """Test that constructor raises error with empty ring ID."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, "", self.TEST_RULE_ID, self.TEST_ISSUER)
        self.assertIn("Ring ID", str(context.exception))

    def test_constructor_with_null_rule_id(self):
        """Test that constructor raises error with null rule ID."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, None, self.TEST_ISSUER)
        self.assertIn("Rule ID", str(context.exception))

    def test_constructor_with_empty_rule_id(self):
        """Test that constructor raises error with empty rule ID."""
        with self.assertRaises(ValueError) as context:
            JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, "", self.TEST_ISSUER)
        self.assertIn("Rule ID", str(context.exception))

    def test_transform_creates_valid_jwe_token(self):
        """Test that transform creates a valid JWE token with correct prefix."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER
        )

        result = formatter.transform(self.TEST_TOKEN)

        # Verify the token has the correct prefix
        self.assertTrue(result.startswith("olt.V1."))

        # Verify it's a valid JWE token (5 parts separated by dots after the prefix)
        jwe_compact = result[len("olt.V1.") :]
        parts = jwe_compact.split(".")
        self.assertEqual(5, len(parts), "JWE compact serialization should have 5 parts")

    def test_transform_with_null_token(self):
        """Test that transform returns None for null token."""
        formatter = JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, None)

        result = formatter.transform(None)
        self.assertIsNone(result)

    def test_transform_with_empty_token(self):
        """Test that transform returns empty string for empty token."""
        formatter = JweMatchTokenFormatter(self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, None)

        result = formatter.transform("")
        self.assertEqual("", result)

    def test_jwe_header_contains_correct_metadata(self):
        """Test that JWE header contains correct metadata."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER
        )

        result = formatter.transform(self.TEST_TOKEN)
        jwe_compact = result[len("olt.V1.") :]

        # Parse the JWE object to inspect the header
        jwe_token = jwe.JWE()
        jwe_token.deserialize(jwe_compact)

        # Get the protected header (already a dict)
        protected_header = jwe_token.jose_header

        # Verify header fields
        self.assertEqual("dir", protected_header["alg"])
        self.assertEqual("A256GCM", protected_header["enc"])
        self.assertEqual("match-token", protected_header["typ"])
        self.assertEqual(self.TEST_RING_ID, protected_header["kid"])

    def test_jwe_payload_contains_correct_data(self):
        """Test that decrypted JWE payload contains correct data."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER
        )

        result = formatter.transform(self.TEST_TOKEN)
        jwe_compact = result[len("olt.V1.") :]

        # Decrypt and parse the payload using the same method as the formatter
        import base64

        key_bytes = self.TEST_ENCRYPTION_KEY.encode("utf-8")
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode("utf-8").rstrip("=")
        jwk_key = jwk.JWK(kty="oct", k=key_b64)

        jwe_token = jwe.JWE()
        jwe_token.deserialize(jwe_compact)
        jwe_token.decrypt(jwk_key)

        payload = json.loads(jwe_token.payload.decode("utf-8"))

        # Verify payload fields
        self.assertEqual(self.TEST_RULE_ID, payload["rlid"])
        self.assertEqual("SHA-256", payload["hash_alg"])
        self.assertEqual("HS256", payload["mac_alg"])
        self.assertEqual([self.TEST_TOKEN], payload["ppid"])
        self.assertEqual(self.TEST_RING_ID, payload["rid"])
        self.assertEqual(self.TEST_ISSUER, payload["iss"])
        self.assertIn("iat", payload)
        self.assertIsInstance(payload["iat"], int)

    def test_default_issuer(self):
        """Test that default issuer is used when None is provided."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY,
            self.TEST_RING_ID,
            self.TEST_RULE_ID,
            None,  # null issuer should default to "org.openlinktoken"
        )

        self.assertEqual("org.openlinktoken", formatter.issuer)

    def test_sha3_suite_metadata_is_embedded_in_olt_v1(self):
        """SHA3 metadata is profile-driven while the token prefix stays olt.V1."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY,
            self.TEST_RING_ID,
            self.TEST_RULE_ID,
            self.TEST_ISSUER,
            CryptoSuite.from_id("suite-sha3-v1"),
        )

        result = formatter.transform(self.TEST_TOKEN)
        token = jwe.JWE()
        token.deserialize(result[len("olt.V1.") :])
        key_bytes = self.TEST_ENCRYPTION_KEY.encode("utf-8")
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode("utf-8").rstrip("=")
        token.decrypt(jwk.JWK(kty="oct", k=key_b64))
        payload = json.loads(token.payload.decode("utf-8"))

        self.assertTrue(result.startswith("olt.V1."))
        self.assertEqual("SHA3-256", payload["hash_alg"])
        self.assertEqual("HS3-256", payload["mac_alg"])

    def test_shake_suite_metadata_is_embedded_in_olt_v1(self):
        """SHAKE metadata records the fixed digest and MAC output lengths."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY,
            self.TEST_RING_ID,
            self.TEST_RULE_ID,
            self.TEST_ISSUER,
            CryptoSuite.from_id("suite-shake-v1"),
        )

        result = formatter.transform(self.TEST_TOKEN)
        token = jwe.JWE()
        token.deserialize(result[len("olt.V1.") :])
        key_bytes = self.TEST_ENCRYPTION_KEY.encode("utf-8")
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode("utf-8").rstrip("=")
        token.decrypt(jwk.JWK(kty="oct", k=key_b64))
        payload = json.loads(token.payload.decode("utf-8"))

        self.assertEqual("SHAKE256-256", payload["hash_alg"])
        self.assertEqual("KMAC256-256", payload["mac_alg"])

    def test_different_tokens_produce_different_outputs(self):
        """Test that different tokens produce different encrypted outputs."""
        formatter = JweMatchTokenFormatter(
            self.TEST_ENCRYPTION_KEY, self.TEST_RING_ID, self.TEST_RULE_ID, self.TEST_ISSUER
        )

        result1 = formatter.transform("token1")
        result2 = formatter.transform("token2")

        self.assertNotEqual(result1, result2)
        # But both should have the same prefix
        self.assertTrue(result1.startswith("olt.V1."))
        self.assertTrue(result2.startswith("olt.V1."))


if __name__ == "__main__":
    unittest.main()

File: lib/java/openlinktoken/src/test/java/org/openlinktoken/tokentransformer/JweMatchTokenFormatterTest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokentransformer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWEObject;
import com.nimbusds.jose.crypto.DirectDecrypter;
import com.nimbusds.jose.jwk.OctetSequenceKey;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.openlinktoken.crypto.CryptoSuite;

/**
 * Unit tests for {@link JweMatchTokenFormatter}.
 */
class JweMatchTokenFormatterTest {

    private static final String TEST_ENCRYPTION_KEY = "12345678901234567890123456789012"; // 32 chars
    private static final String TEST_RING_ID = "test-ring-2026";
    private static final String TEST_RULE_ID = "T1";
    private static final String TEST_TOKEN = "dGVzdC10b2tlbi1wcGlk"; // base64-encoded test token

    @Test
    void testConstructorWithValidParameters() throws JOSEException {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                "test.issuer");
        assertNotNull(formatter);
    }

    @Test
    void testConstructorWithRaw32ByteKey() throws JOSEException {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY.getBytes(StandardCharsets.UTF_8),
                TEST_RING_ID,
                TEST_RULE_ID,
                "test.issuer");
        assertNotNull(formatter);
    }

    @Test
    void testConstructorWithNullEncryptionKey() {
        assertThrows(IllegalArgumentException.class, () -> {
            new JweMatchTokenFormatter((String) null, TEST_RING_ID, TEST_RULE_ID, "test.issuer");
        });
    }

    @Test
    void testConstructorWithInvalidKeyLength() {
        assertThrows(IllegalArgumentException.class, () -> {
            new JweMatchTokenFormatter("short", TEST_RING_ID, TEST_RULE_ID, "test.issuer");
        });
    }

    @Test
    void testConstructorWithNonAscii32CharacterKey() {
        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> {
            new JweMatchTokenFormatter("é".repeat(32), TEST_RING_ID, TEST_RULE_ID, "test.issuer");
        });
        assertEquals("Encryption key must be exactly 32 bytes (256 bits)", exception.getMessage());
    }

    @Test
    void testConstructorWithNullRingId() {
        assertThrows(IllegalArgumentException.class, () -> {
            new JweMatchTokenFormatter(TEST_ENCRYPTION_KEY, null, TEST_RULE_ID, "test.issuer");
        });
    }

    @Test
    void testConstructorWithNullRuleId() {
        assertThrows(IllegalArgumentException.class, () -> {
            new JweMatchTokenFormatter(TEST_ENCRYPTION_KEY, TEST_RING_ID, null, "test.issuer");
        });
    }

    @Test
    void testTransformCreatesValidJweToken() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                "test.issuer");

        String result = formatter.transform(TEST_TOKEN);

        // Verify the token has the correct prefix
        assertTrue(result.startsWith("olt.V1."));

        // Verify it's a valid JWE token (5 parts separated by dots after the prefix)
        String jweCompact = result.substring("olt.V1.".length());
        String[] parts = jweCompact.split("\\.");
        assertEquals(5, parts.length, "JWE compact serialization should have 5 parts");
    }

    @Test
    void testTransformWithNullToken() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                null);

        String result = formatter.transform(null);
        assertNull(result);
    }

    @Test
    void testTransformWithEmptyToken() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                null);

        String result = formatter.transform("");
        assertEquals("", result);
    }

    @Test
    void testJweHeaderContainsCorrectMetadata() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                "test.issuer");

        String result = formatter.transform(TEST_TOKEN);
        String jweCompact = result.substring("olt.V1.".length());

        // Parse the JWE object to inspect the header
        JWEObject jweObject = JWEObject.parse(jweCompact);

        // Verify header fields
        assertEquals("dir", jweObject.getHeader().getAlgorithm().getName());
        assertEquals("A256GCM", jweObject.getHeader().getEncryptionMethod().getName());
        assertEquals("match-token", jweObject.getHeader().getType().getType());
        assertEquals(TEST_RING_ID, jweObject.getHeader().getKeyID());
    }

    @Test
    void testShakeSuiteMetadataIsEmbeddedInOltV1() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY.getBytes(StandardCharsets.UTF_8),
                TEST_RING_ID,
                TEST_RULE_ID,
                "test.issuer",
                CryptoSuite.fromId("suite-pq-shake-v1"));

        String result = formatter.transform(TEST_TOKEN);
        JWEObject jweObject = JWEObject.parse(result.substring("olt.V1.".length()));
        jweObject.decrypt(new DirectDecrypter(
                new OctetSequenceKey.Builder(TEST_ENCRYPTION_KEY.getBytes(StandardCharsets.UTF_8)).build()));
        Map<String, Object> payload = jweObject.getPayload().toJSONObject();

        assertEquals("SHAKE256-256", payload.get("hash_alg"));
        assertEquals("KMAC256-256", payload.get("mac_alg"));
    }

    @Test
    void testDefaultIssuer() throws Exception {
        JweMatchTokenFormatter formatter = new JweMatchTokenFormatter(
                TEST_ENCRYPTION_KEY,
                TEST_RING_ID,
                TEST_RULE_ID,
                null // null issuer should default to "org.openlinktoken"
        );

        assertNotNull(formatter);
        // The default issuer is set internally and will be verified in the decrypted payload
    }
}
